File: 1530.cpp
#include <map>

struct Solution {
    using Dict = map<int, int>;

    Dict dfs(TreeNode *node, int distance, int &count){
        Dict d;
        if (node == nullptr){
            return d;
        }
        if (node->left == nullptr && node->right == nullptr){
            d[0] = 1;
            return d;
        }
        Dict ld = dfs(node->left, distance, count);
        Dict rd = dfs(node->right, distance, count);
        for (auto &lpr : ld){
            for (auto &rpr : rd){
                if (lpr.first + rpr.first + 2 > distance){
                    break;
                }
                count += lpr.second * rpr.second;
            }
        }
        for (auto &pr : ld){
            if (pr.first >= distance - 2){
                break;
            }
            d[pr.first + 1] += pr.second;
        }
        for (auto &pr : rd){
            if (pr.first >= distance - 2){
                break;
            }
            d[pr.first + 1] += pr.second;
        }
        return d;
    }

    inline int countPairs(TreeNode* root, int distance) {
        int count = 0;
        this->dfs(root, distance, count);
        return count;
    }
};

File: sol/1530.py
from collections import defaultdict

class Solution:
    def countPairs(self, root: TreeNode, distance: int) -> int:
        count = 0
        def dfs(node):
            d = defaultdict(int)
            if node is None:
                return d
            if node.left is None and node.right is None:
                d[0] = 1
                return d
            ld = dfs(node.left)
            rd = dfs(node.right)
            lk = sorted(ld.keys())
            rk = sorted(rd.keys())
            nonlocal count
            for a in lk:
                for b in rk:
                    if a + b + 2 > distance:
                        break
                    count += ld[a] * rd[b]
            for a in lk:
                if a + 2 >= distance:
                    break
                d[a + 1] += ld[a]
            for b in rk:
                if b + 2 >= distance:
                    break
                d[b + 1] += rd[b]
            return d
        dfs(root)
        return count
